src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -821,12 +822,21 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             .iter()
             .filter_map(|expr| tables.node_type_opt(expr.hir_id))
             .map(|ty| self.resolve_vars_if_possible(&ty));
-        let (last_ty, all_returns_have_same_type) = ret_types.clone().fold(
-            (None, true),
-            |(last_ty, mut same): (std::option::Option<Ty<'_>>, bool), ty| {
+        let (last_ty, all_returns_have_same_type, never_count) = ret_types.clone().fold(
```

Comment:
Why are we counting `!`-types instead of checking whether there is one or not? We don't seem to be using the actual count anywhere.

Reply:
That's fair. Originally this was a count of how many return paths there were. Right now it should be a boolean signaling whether _any_ return paths were found. If there are none, we are dealing with either an empty body or a body that returns `!`. In either case most likely we have in-progress code so we need to "guess" what to suggest, so we suggest everything ("return a type! or Box<dyn Trait>! or impl Trait! :D").